### ISSUE TITLE
Rails 5.1.x does blow up when queries are executed within model class definition

### ISSUE THREAD

**Issue by jarmo** [rails/rails#29419]
When an autoload for any model class is triggered and it has a querying code in its class definition, then starting from Rails 5.1.0 it does not work anymore.

It happened to me in real life when running tests in CI by using `devise_for :users` helper (from Devise library) in routes, which in turn does `"user".contantize` and this blows up since ActiveRecord tries to load `users` table, which does not yet exist (every time tests are ran, it starts from scratch, as usual). I understand that having a query in model class definition might not be a good practice (I'm using it to load administrators and cache them since it is a "constant"), but it was working perfectly in Rails < 5.1.0 (this can be verified, by changing `RAILS_VERSION` constant to 5.0.3 within test script provided below and running it).

In addition to the crash itself - why are routes even loaded when trying to run any **db** specific tasks?

### Steps to reproduce

Here's a code, which can be ran in your terminal to reproduce the problem:
```
# setup
RAILS_VERSION="5.1.1"

gem install rails -v $RAILS_VERSION && \
  rails _${RAILS_VERSION}_ new rails-${RAILS_VERSION}-test && \
  cd rails-${RAILS_VERSION}-test && \
  rails generate model User && \
  rails db:migrate && \
  echo 'class User < ActiveRecord::Base; FOO = where(id: [1]); end' > app/models/user.rb && \
  echo 'Rails.application.routes.draw { User }' > config/routes.rb

# test
RAILS_ENV=test rails db:drop && \
  RAILS_ENV=test rails db:create db:schema:load --trace

# cleanup
cd .. && \
  rm -rf rails-${RAILS_VERSION}-test
```

### Expected behavior

Expected behavior is no crashes or a better error message since behavior has changed between older versions of Rails.

### Actual behavior

It crashes with an error message like this:
```
rails aborted!
ActiveRecord::StatementInvalid: Could not find table 'users'
/Users/jarmo/.gem/ruby/2.3.1/gems/activerecord-5.1.1/lib/active_record/connection_adapters/sqlite3_adapter.rb:417:in `table_structure'
/Users/jarmo/.gem/ruby/2.3.1/gems/activerecord-5.1.1/lib/active_record/connection_adapters/abstract_adapter.rb:166:in `columns'
/Users/jarmo/.gem/ruby/2.3.1/gems/activerecord-5.1.1/lib/active_record/connection_adapters/schema_cache.rb:67:in `columns'
/Users/jarmo/.gem/ruby/2.3.1/gems/activerecord-5.1.1/lib/active_record/connection_adapters/schema_cache.rb:73:in `columns_hash'
/Users/jarmo/.gem/ruby/2.3.1/gems/activerecord-5.1.1/lib/active_record/model_schema.rb:451:in `load_schema!'
/Users/jarmo/.gem/ruby/2.3.1/gems/activerecord-5.1.1/lib/active_record/attributes.rb:233:in `load_schema!'
/Users/jarmo/.gem/ruby/2.3.1/gems/activerecord-5.1.1/lib/active_record/attribute_decorators.rb:50:in `load_schema!'
/Users/jarmo/.gem/ruby/2.3.1/gems/activerecord-5.1.1/lib/active_record/model_schema.rb:446:in `load_schema'
...
```

### System configuration
**Rails version**: 5.1.1

**Ruby version**: 2.3.1p112

**Comment by jarmo** [comment]
Another question in addition to the "why are routes loaded at all":
* Why is that query performed in the first place? Shouldn't it be lazy like it has been so far?

**Comment by nthj** [comment]
I was able to reproduce and pasted my [full output to this gist](https://gist.github.com/nthj/3be9af5be0de314fdbe074be12b01079).

Anecdotally, my experience has been for as long as I can remember that I need `if table_exists?`-type checks, or I get weird behavior with `db:schema:load` or `db:migrate:reset`. I checked the changelog and the history for the files in the stack trace and I didn't see anything that jumped out as a potential cause. I'm curious if the 5.0 behavior was only ever accidentally supported.

**Comment by jarmo** [comment]
Yes, it's possible that it has been accidentally working thus far (prior 5.1, even 4.x worked like this), but there's still questions remaining as stated above:

1) why are routes loaded in the first place, when `db` tasks are used? Is it because of `environment` task as a dependency?
2) Why is that query executed at all? Isn't `where` and other query statements lazily evaluated anymore?

And one additional question - what is the Rails way of doing things like this:

```ruby
class User < ActiveRecord::Base
  ADMINISTRATORS = where(admin: 1) # or whatever
end
```

Yes, I understand that scopes exist, but the solution above is better since:
1) Using a constant `ADMINISTRATORS` implies that administrators are not changed during runtime (and they aren't in my case) - in other words, it is a constant and using scope here wouldn't make sense;
2) Query is executed only once and its result is cached "forever" (not taking into account that Rails can be ran on multiple nodes);

**Comment by meinac** [comment]
What I recommend you is, just define a singleton method and use memoization. By doing this you can postpone the query call to first time you call it and cache it forever.

```ruby
class User < ActiveRecord::Base
  def self.administrators
    @administrators ||= where(admin: 1).to_a
  end
end
```

Note that, you need to call a **kicker** method for relation like `to_a`, otherwise you will hit to the database whenever you call this method.

Oh, and also I'll try to find the reason of this behaviour change let you know.
